### ISSUE TITLE
[rewrite] UpdateStyle regression in commit 115502

### ISSUE THREAD

**Issue by sebastiansandqvist** [lhorie/mithril.js#1206]
### Description:

Upgraded from commit #fe630b to commit #115502 in the rewrite, and got the following error on code that previously worked just fine:

`Uncaught TypeError: Cannot set property 'style' of undefined`

It happens in the `updateStyle` function when trying to access `element.style`.
https://github.com/lhorie/mithril.js/blob/1155027c9125f9d157318d0c2d7583b90e2e1832/render/render.js#L444

In the debugger, I have the following values for each of the parameters:

`element = undefined`
`old = { width: "12px", height: "12px" }`
`style = { width: "12px", height: "12px" }`

I tried the latest commit as well, and it is broken, too. #115502 is just the earliest commit I could find that fails.

**Comment by pygy** [comment]
The stack trace itself may help, since `setAttr` can be called from several places. The value of  `vnode` in the `setAttr` frame may also help you pinpoint the troublesome item.

**Comment by isiahmeadows** [comment]
Stack trace?

**Comment by isiahmeadows** [comment]
Could you read the stack trace and walk up it, listing values for each one? That would be most useful.

**Comment by sebastiansandqvist** [comment]
It likely has something to do with this being an svg.

Within `setAttr`, the parameters are:

```js
vnode = {
	attrs: Object,
	children: Array[1],
	dom: undefined,
	domSize: undefined,
	events: undefined,
	instance: undefined,
	key: undefined,
	state: Object,
	tag: "svg",
	text: undefined,
	__proto__: Object
},
key = "style",
old = {width: "12px", height: "12px" },
value = {width: "12px", height: "12px" },
ns = "http://www.w3.org/2000/svg"
```

Because `element` is set to `vnode.dom`, it is undefined when passed to `updateStyle`

The call stack is pretty large, but here it is:

```
updateStyle	@	mithril-52c3233….js:740
setAttr	@	mithril-52c3233….js:690
updateAttrs	@	mithril-52c3233….js:714
updateElement	@	mithril-52c3233….js:542
updateNode	@	mithril-52c3233….js:494
updateComponent	@	mithril-52c3233….js:557
updateNode	@	mithril-52c3233….js:497
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateComponent	@	mithril-52c3233….js:557
updateNode	@	mithril-52c3233….js:497
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateComponent	@	mithril-52c3233….js:557
updateNode	@	mithril-52c3233….js:497
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateComponent	@	mithril-52c3233….js:557
updateNode	@	mithril-52c3233….js:497
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateNodes	@	mithril-52c3233….js:420
updateElement	@	mithril-52c3233….js:549
updateNode	@	mithril-52c3233….js:494
updateComponent	@	mithril-52c3233….js:557
updateNode	@	mithril-52c3233….js:497
updateComponent	@	mithril-52c3233….js:557
updateNode	@	mithril-52c3233….js:497
updateComponent	@	mithril-52c3233….js:557
updateNode	@	mithril-52c3233….js:497
updateNodes	@	mithril-52c3233….js:420
render	@	mithril-52c3233….js:816
render	@	mithril-52c3233….js:1094
router.defineRoutes.router.setPath.replace	@	mithril-52c3233….js:1098
(anonymous function)	@	mithril-52c3233….js:1468
(anonymous function)	@	mithril-52c3233….js:1462
onNextTick	@	app-a2264b1….js:4448
Item.run	@	app-a2264b1….js:6117
drainQueue	@	app-a2264b1….js:6087
setTimeout (async)		
runTimeout	@	app-a2264b1….js:6042
process.nextTick	@	app-a2264b1….js:6107
exports.setImmediate	@	app-a2264b1….js:4441
resolveRoute	@	mithril-52c3233….js:1457
(anonymous function)	@	mithril-52c3233….js:1134
publish	@	mithril-52c3233….js:1066
(anonymous function)	@	app-a2264b1….js:5098
(anonymous function)	@	app-a2264b1….js:63
setTimeout (async)		
(anonymous function)	@	app-a2264b1….js:62
(anonymous function)	@	mithril-52c3233….js:1003
(anonymous function)	@	mithril-52c3233….js:958
updateDependency	@	mithril-52c3233….js:912
initDependency	@	mithril-52c3233….js:979
combine	@	mithril-52c3233….js:955
map	@	mithril-52c3233….js:1003
oninit	@	app-a2264b1….js:5102
initLifecycle	@	mithril-52c3233….js:776
createComponent	@	mithril-52c3233….js:392
createNode	@	mithril-52c3233….js:328
updateNode	@	mithril-52c3233….js:501
updateNodes	@	mithril-52c3233….js:420
render	@	mithril-52c3233….js:816
render	@	mithril-52c3233….js:1094
router.defineRoutes.router.setPath.replace	@	mithril-52c3233….js:1098
(anonymous function)	@	mithril-52c3233….js:1468
(anonymous function)	@	mithril-52c3233….js:1462
onNextTick	@	app-a2264b1….js:4448
Item.run	@	app-a2264b1….js:6117
drainQueue	@	app-a2264b1….js:6087
setTimeout (async)		
runTimeout	@	app-a2264b1….js:6042
process.nextTick	@	app-a2264b1….js:6107
exports.setImmediate	@	app-a2264b1….js:4441
resolveRoute	@	mithril-52c3233….js:1457
(anonymous function)	@	mithril-52c3233….js:1134
publish	@	mithril-52c3233….js:1066
(anonymous function)	@	app-a2264b1….js:63
setTimeout (async)		
(anonymous function)	@	app-a2264b1….js:62
(anonymous function)	@	mithril-52c3233….js:1003
(anonymous function)	@	mithril-52c3233….js:958
updateDependency	@	mithril-52c3233….js:912
updateStream	@	mithril-52c3233….js:881
stream	@	mithril-52c3233….js:837
(anonymous function)	@	firebase-a01fc7a….js:44
(anonymous function)	@	firebase-a01fc7a….js:46
﻿
```

**Comment by isiahmeadows** [comment]
Also, what code would repro this?

**Comment by sebastiansandqvist** [comment]
It appears to be an icon within a cached component:

```js
		state.input = tempQuery.map(function(query) {
			return (
				m('div',
					m(Icon, { name: 'search', className: 'fill-slate', size: 12 }),
					m(Input, {
						className: 'noborder market-search',
						placeholder: 'Search listings',
						maxlength: '30',
						oninput(e) { tempQuery(e.target.value); },
						value: query
					}),
					buttons()
				)
			);
		});
```


Where `Icon` is defined as:

```js
const Icon = {
	view({ attrs }) {
		const className = attrs.className || '';
		const size = attrs.size;
		const style = attrs.style || (size ? { width: `${size}px`, height: `${size}px` } : {});
		const viewBox = attrs.viewBox || '0 0 20 20';
		return m('svg.Icon', { className, style, viewBox }, icons[attrs.name] );
	}
};
```

And the reference to `icons[attrs.name]` accesses something of the form
```js
m('path', { d: /* svg path data */ )
```

**Comment by sebastiansandqvist** [comment]
Just so there is a reference here, that commit fixed this issue: https://github.com/lhorie/mithril.js/issues/1163

**Comment by isiahmeadows** [comment]
@sebastiansandqvist Could you make that repro any smaller (and self-sufficient in terms of JS)?

**Comment by pygy** [comment]
Does the icon ever appear on screen?

**Comment by sebastiansandqvist** [comment]
Here's a repro repo: https://github.com/sebastiansandqvist/mithril-bug-repro
The good stuff is in src.js

For this issue to arise:
- Inner component must have a `style` attr
- Outer component must be memoized
- Immediate async redraw must occur in oninit

@pygy The component causing the issue does appear on the screen, but it turns out it does not need to be an `svg`. Logging `vnode.dom` works just as you would expect it to in the `oncreate`.

**Comment by pygy** [comment]
@sebastiansandqvist Thanks for the test case. This will hopefully be fixed when #1224 lands.

**Comment by sebastiansandqvist** [comment]
Looks like it's been merged as of commit 50a80e55. Thanks @pygy and @lhorie.
